lib/srv/monitor.go

```diff
@@ -145,84 +170,125 @@ func (w *Monitor) Start() {
+			if err := w.emitAuditEvent(reason); err != nil {
 				w.Entry.WithError(err).Warn("Failed to emit audit event.")
 			}
-			w.Entry.Debugf("Disconnecting client: %v", event.Reason)
-			w.Conn.Close()
+			w.Entry.Debugf("Disconnecting client: %v", reason)
+			if err := w.Conn.Close(); err != nil {
+				w.Entry.WithError(err).Error("Failed to close connection.")
+			}
 			return
+
+		// Idle timeout.
 		case <-idleTime:
 			now := w.Clock.Now().UTC()
 			clientLastActive := w.Tracker.GetClientLastActive()
 			if now.Sub(clientLastActive) >= w.ClientIdleTimeout {
-				event := &apievents.ClientDisconnect{
-					Metadata: apievents.Metadata{
-						Type: events.ClientDisconnectEvent,
-						Code: events.ClientDisconnectCode,
-					},
-					UserMetadata: apievents.UserMetadata{
-						Login: w.Login,
-						User:  w.TeleportUser,
-					},
-					ConnectionMetadata: apievents.ConnectionMetadata{
-						LocalAddr:  w.Conn.LocalAddr().String(),
-						RemoteAddr: w.Conn.RemoteAddr().String(),
-					},
-					ServerMetadata: apievents.ServerMetadata{
-						ServerID: w.ServerID,
-					},
-				}
-				if clientLastActive.IsZero() {
-					event.Reason = "client reported no activity"
-				} else {
-					event.Reason = fmt.Sprintf("client is idle for %v, exceeded idle timeout of %v",
+				reason := "client reported no activity"
+				if !clientLastActive.IsZero() {
+					reason = fmt.Sprintf("client is idle for %v, exceeded idle timeout of %v",
 						now.Sub(clientLastActive), w.ClientIdleTimeout)
 				}
-				w.Entry.Debugf("Disconnecting client: %v", event.Reason)
-
+				if err := w.emitAuditEvent(reason); err != nil {
+					w.Entry.WithError(err).Warn("Failed to emit audit event.")
+				}
 				if w.MessageWriter != nil && w.IdleTimeoutMessage != "" {
 					if _, err := w.MessageWriter.WriteString(w.IdleTimeoutMessage); err != nil {
 						w.Entry.WithError(err).Warn("Failed to send idle timeout message.")
 					}
 				}
-				w.Conn.Close()
-
-				if err := w.Emitter.EmitAuditEvent(w.Context, event); err != nil {
-					w.Entry.WithError(err).Warn("Failed to emit audit event.")
+				w.Entry.Debugf("Disconnecting client: %v", reason)
+				if err := w.Conn.Close(); err != nil {
+					w.Entry.WithError(err).Error("Failed to close connection.")
 				}
 				return
 			}
 			w.Entry.Debugf("Next check in %v", w.ClientIdleTimeout-now.Sub(clientLastActive))
 			idleTime = w.Clock.After(w.ClientIdleTimeout - now.Sub(clientLastActive))
+
+		// Lock in force.
+		case lockEvent := <-lockWatch.Events():
+			lock, err := getLock(lockEvent)
+			if err != nil {
+				w.Entry.WithError(err).Warnf("Failed to extract lock from event %v.", lockEvent)
+				continue Loop
+			}
+			w.handleLockInForce(lock)
+			return
+
+		case <-lockWatch.Done():
+			w.Entry.WithError(lockWatch.Error()).Warn("Lock watcher subscription was closed.")
+			if w.DisconnectExpiredCert.IsZero() && w.ClientIdleTimeout == 0 {
+				return
+			}
+			continue Loop
```

Comment:
if `lockWatch` done() channel was close and Idle and Disconnect timeouts were set we will re-run the loop and create a spin-lock because on next iteration of `select` `lockWatch.Done()` statement  channel the lockWatch.Done channel will be also closed.

Reply:
You're right. This case should actually never trigger, I'll remove it entirely.